### PR TITLE
Unbounded fetch

### PR DESCRIPTION
**Pull request by jsnoble**
- all fetchers will now try to pull all records in their given slices to work with active indicies
- spaces_reader_api and elasticsearch_reader_api will fetch window_size from their given index assigned.
- updated api schemas
- updated docs

### CODE REVIEW

**Review by macgyver603 — APPROVED** [review]
LGTM

File: asset/src/elasticsearch_reader_api/reader.ts [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+} from '@terascope/job-components';
+import { CountParams, SearchParams } from 'elasticsearch';
+import { buildQuery } from '../elasticsearch_reader/elasticsearch_date_range/helpers';
+import { ESReaderOptions, SlicerDateResults } from '../elasticsearch_reader/interfaces';
+
+export default class APIReader {
+    readonly config: ESReaderOptions;
+    logger: Logger;
+    _baseClient: AnyObject;
+    client: elasticAPI.Client;
```
Comment (peterdemartini) [inline asset/src/elasticsearch_reader_api/reader.ts:18]: We shouldn't be exposing the underlying client.

```suggestion
    protected readonly client: elasticAPI.Client;
```

File: asset/src/elasticsearch_reader_api/reader.ts [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+    toNumber,
+} from '@terascope/job-components';
+import { CountParams, SearchParams } from 'elasticsearch';
+import { buildQuery } from '../elasticsearch_reader/elasticsearch_date_range/helpers';
+import { ESReaderOptions, SlicerDateResults } from '../elasticsearch_reader/interfaces';
+
+export default class APIReader {
+    readonly config: ESReaderOptions;
+    logger: Logger;
+    _baseClient: AnyObject;
```
Comment (peterdemartini) [inline asset/src/elasticsearch_reader_api/reader.ts:17]:
```suggestion
    private _baseClient: AnyObject;
```

File: asset/src/elasticsearch_reader_api/reader.ts [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+    client: elasticAPI.Client;
+    private hasDefaultQueries = false;
+    private windowSize: undefined | number = undefined;
+
+    constructor(config: ESReaderOptions, client: AnyObject, logger: Logger) {
```
Comment (peterdemartini) [inline asset/src/elasticsearch_reader_api/reader.ts:22]: Why is the client an `AnyObject`?

**Review by peterdemartini — CHANGES_REQUESTED** [review]

File: asset/src/elasticsearch_reader_api/reader.ts [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+        }
+    }
+
+    async count(queryParams: Partial<SlicerDateResults> = {}): Promise<number> {
+        this.validate(queryParams);
+        const query = buildQuery(this.config, queryParams);
+        return this.client.count(query as CountParams);
+    }
+
+    async fetch(queryParams: Partial<SlicerDateResults> = {}): Promise<DataEntity[]> {
```
[the remark this answers is not in the document]
Comment (jsnoble) [reply asset/src/elasticsearch_reader_api/reader.ts:57]: this method is fetching records for the given slice, naming it something like search felt wrong as we restrict and limit searching to only date range and wildcard searches compared to elasticsearch search function. As for the documentation comment, what do you think is lacking in  ./docs/apis/elasticsearch_reader_api to explain how to use this?

File: asset/src/elasticsearch_reader/elasticsearch_date_range/slicer.ts [hunk 5]
```diff
@@ -96,7 +95,7 @@ export default class DateSlicer extends ParallelSlicer<ESDateConfig> {
         }
 
         // using this query to catch potential errors even if a date is given already
-        const results = await this.api.search(query);
+        const results = await this.api._searchRequest(query);
```
[the remark this answers is not in the document]
Comment (jsnoble) [reply asset/src/elasticsearch_reader/elasticsearch_date_range/slicer.ts:98]: what are you envision as an abstraction? this is the backdoor we talked about last friday. This is the unsupported api that will most likely change in the future, but allows them to make any search query they want.

File: asset/src/elasticsearch_reader_api/reader.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (peterdemartini) [reply asset/src/elasticsearch_reader_api/reader.ts:60]: But `fetch` is very generic for something that takes a `SlicerDateResults` as params? I think naming it something like `fetchRange` would be more appropriate? 

It is super useful to have jsdoc's while developing and it is closer to the source of truth (meaning it is easier to keep in sync when making changes), the docs `./docs/apis/elasticsearch_reader_api` should probably be auto-generated from the jsdocs, but for now they should be copied and pasted. (This how most of our libraries operate and it is fairly common among other packages).

File: asset/src/elasticsearch_reader_api/reader.ts [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+        try {
+            const settings = await this._baseClient.indices.getSettings(query);
+
+            const defaultPath = settings[index].defaults[window];
+            const configPath = settings[index].settings[window];
+
+            if (defaultPath) this.windowSize = toNumber(defaultPath);
+            if (configPath) this.windowSize = toNumber(configPath);
+        } catch (_err) {
+            this.logger.warn(`index: ${this.config.index} is not yet created`);
```
Comment (peterdemartini) [inline asset/src/elasticsearch_reader_api/reader.ts:99]: Not every error throw is going to be a "Index Doesn't Exist" error, we should be more explicit about this check.

File: asset/src/elasticsearch_reader/elasticsearch_date_range/slicer.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (peterdemartini) [reply asset/src/elasticsearch_reader/elasticsearch_date_range/slicer.ts:98]: I think you misinterpreted me, I said *we* shouldn't be using it the lower level API but we can't expose once external developers. I think the whole `getIndexDate` function should be the abstraction.

**Comment by jsnoble** [comment]
**Elasticsearch API Overview**
### fetch
```(query: ElasticsearchSliceQuery) => Promise<DataEntities[]>```
This will perform an date range or wildcard query to elasticsearch and return the results of the query.

parameters:
- query: an slice query object
  -  start: string, must be paired with end to do a date range query.
  -  end: string, must be paired with start to do a date range query.
  -  wildcard: { field: string, value: string }, an elasticsearch wildcard query on string values. The value needs to be formatted in `key*`,please reference examples below.
  -  key: string, only used for _uid queries on elasticsearch v5 or older. The key need to be specified as `docType#key*` format, please reference examples below.

```js
const dateRangeQuery = {
    start: '2019-04-26T15:00:23.201Z',
    end: '2019-04-26T15:20:23.220Z',
};

let results = await api.fetch(dateRangeQuery);
results === [ { some: 'record', created: '2019-04-26T15:10:23.220Z' }]

const oldUIDQuery = {
   key:  `events#ba*`
};

let results = await api.fetch(oldUIDQuery);
results === [ { some: 'record', created: '2019-04-26T15:10:23.220Z' }]

const wildcardQuery = {
    field: 'uuid,
    value: 'afe1*',
};

let results = await api.fetch(wildcardQuery);
results === [ { some: 'record', uuid: 'afe18550-0081-453f-9e80-93a90782a5bd' }]
```


### count
```(query: ElasticsearchQuery) => Promise<number>```
This will perform an count query and return the number of records in that are in the range of the query.

parameters:
- query: an slice query object
  -  start: string, must be paired with end to do a date range query.
  -  end: string, must be paired with start to do a date range query.
  -  wildcard: { field: string, value: string }, an elasticsearch wildcard query on string values. The value needs to be formatted in `key*`,please reference examples below.
  -  key: string, only used for _uid queries on elasticsearch v5 or older. The key need to be specified as `docType#key*` format, please reference examples below.

```js
const dateRangeQuery = {
    start: '2019-04-26T15:00:23.201Z',
    end: '2019-04-26T15:20:23.220Z',
};

let results = await api.count(dateRangeQuery);
results === 122

const oldUIDQuery = {
   key:  `events#ba*`
};

let results = await api.fetch(oldUIDQuery);
results === 3552

const wildcardQuery = {
    field: 'uuid,
    value: 'afe1*',
};

let results = await api.fetch(wildcardQuery);
results === 1232
```

### _searchRequest
```(query: ElasticsearchSearchParams) => Promise<DataEntities[]>```
This will allow you to pass in a whole elasticsearch query object to make custom queries. THIS IS AN UNSUPPORTED ESCAPE HATCH. Please do not overly rely on this as this is an internal api and will most likely change.

parameters:
- query: an elasticsearch query object

```js
const query: {
    q: '(test:query OR other:thing AND bytes:>=2000)',
    size: 100,
    fields: 'foo,bar,date'
};

const results = await api._searchRequest(query);

api.version === 6
```

### version
```() => number```
This returns the major elasticsearch version that this client is connected to

```js
api.version === 6
```

### verifyIndex
```() => Promise<void>```
This check if the index exists and throw otherwise, this will also log the window_size of that given index.

```js
try {
    await api.verifyIndex()
} catch(err) {
    // handle error or create index because it does not exists or is not available
}
```

### determineSliceInterval
```(interval: string|duration, dateRange?: DateSegments) => Promise<[number, "time unit"]>```

This is a helper api that will determine how big the slice interval should be for your date_slicer, this is already handled in `makeDateSlicer` for you. If you are constructing your own slicer you can use this to help determine the appropriate size. If interval is set to `auto`, dateRange must be provided and the function will calculate and interval for you. If passed in a duration, it will parse it for you in the format listed by the `time_resolution` configuration.

parameters:
- interval: a duration string (ie. `5min`, `30s`, `750ms`), or it may be set to `auto`
- dateRange: optional, only used when interval is set to `auto`,
  - start: a start date
  - limit: the end date


```js
const apiConfig = {
    time_resolution: 'ms'
}

const interval = await api.determineSliceInterval('2s');
interval === [2000, 'ms'];


const dateRange = {
    start: '2019-04-26T15:00:23.201Z',
    limit: '2019-04-26T15:20:23.220Z',
};

const interval = await api.determineSliceInterval('auto', dateRange);
interval === [2763533, 'ms'];
```

### makeDateSlicer
```(args: DateSlicerArgs) => Promise<SlicerFn>```

This function will generate a slicer which is the elasticsearch_reader slicer core component. You can use this to generate slice chunks for your reader.

parameters:
- args: an slice query object
  -  lifecycle: must be set to `once` or `persistent`.
  -  slicerID: the numerical id of the slicer.
  -  numOfSlicers: the number of total slicers that will be generated
  -  recoveryData?: the `SlicerRecoveryData` for the job
  -  windowState?: the windowState controller, *only needed in persistent mode*
  -  startTime?: the start time of the operation, *only needed in persistent mode*
  -  hook?: an optional async callback that you can provide to handle the updates that the slicer creates (start/end times of the index, the interval for the job etc)

Once mode configuration
```js
const lifecycle = 'once';
const numOfSlicers = 3;
const slicerID = 0;
const recoveryData = [];

const hook = async (params) => {
    // params: { interval: [number, "time unit"], start: Date, end: Date}
  console.log(`Final job configuration ${JSON.stringify(params)}`)
};

const slicerConfig: DateSlicerArgs = {
   lifecycle,
   numOfSlicers,
   slicerID,
   recoveryData,
   hook,
};

const slicer = await this.api.makeDateSlicer(slicerConfig);

const results = await slicer();
results === {
    start: '2019-04-26T15:00:23.201Z',
    end: '2019-04-26T16:00:23.201Z',
    limit: '2019-04-26T15:20:23.220Z',
    count: 3467
}
```

Persistent mode configuration
```js
const lifecycle = 'persistent';
const numOfSlicers = 3;
const slicerID = 0;
const recoveryData = [];
const startTime = new Date().toISOString();
const windowState = this.api.makeWindowState(slicers);

const slicerConfig: DateSlicerArgs = {
   lifecycle,
   numOfSlicers,
   slicerID,
   recoveryData,
   startTime,
   windowState
};

const slicer = await this.api.makeDateSlicer(slicerConfig);

const results = await slicer();

results === {
    start: '2019-04-26T15:00:23.201Z',
    end: '2019-04-26T16:00:23.201Z',
    limit: '2019-04-26T15:20:23.220Z',
    count: 3467
}
```

### makeIDSlicer
```(args: IDSlicerArgs) => Promise<SlicerFn>```

This function will generate a slicer which is the id_reader slicer core component. You can use this to generate slice chunks for your reader.

parameters:
- args: an slice query object
  -  lifecycle: must be set to `once` or `persistent`.
  -  slicerID: the numerical id of the slicer.
  -  numOfSlicers: the number of total slicers that will be generated
  -  recoveryData?: the `SlicerRecoveryData` for the job
  -  key_type: The type of id used in index or the chars contained in the field'
  -  key_range?: if provided, slicer will only recurse on these given keys
  -  starting_key_depth: the length of the key generated before attempting to count the records in the index


```js
const lifecycle = 'once';
const numOfSlicers = 3;
const slicerID = 0;
const recoveryData = [];
const key_type = 'hexadecimal';
const starting_key_depth = 2;

const args: IDSlicerArgs = {
    lifecycle,
    numOfSlicers,
    slicerID,
    recoveryData,
    key_type,
    starting_key_depth,
};

const slicer = await this.api.makeIDSlicer(args);

const results = await slicer();
results ===  { key: 'a0*', count: 5 }
```
### determineDateRanges
```(args: IDSlicerArgs) => Promise<{ start: moment.Moment; limit: moment.Moment; }>```

This is a helper api to parse the start/end dates set on the apiConfig. If no start or end is specified, it will query against the index to find the earliest and latest record and produce dates to include them.

```js

const apiConfig = {
    start: '2019-04-26T15:00:23.201Z',
    index: 'some_index'
}

const results = await api.determineDateRanges()

results ==== {
    start: moment('2019-04-26T15:00:23.201Z'),
    limit: moment('2019-04-26T15:20:23.220Z'),
}

```

### getWindowSize
```() => Promise<number | null>```

this is a helper method to get the `index.max_result_window` size setting configured on the index. This is useful to determine how large a slice is permitted.

```js
const size = await api.getWindowSize();

size === 100000
```

### makeWindowState
```(numOfSlicers: number) => WindowState```
This is a helper api that will return a synchronization window to coordinate slicer date range progression. THIS IS ONLY NEEDED FOR A DATE SLICER IN *PERSISTENT* MODE. This needs to be made once at the top level of the processor/slicer and passed in each time `makeDateSlicer` is called.

```js
const windowState = await api.makeWindowState();
```

**Review by peterdemartini — APPROVED** [review]
Looks much better, the reader api is still missing jsdocs but I am not going to block it because of it. We **really** do need to be better about using jsdoc in our exposed APIs

**Comment by kstaken** [comment]
API looks good in general. docs have some detail issues copy and pasted in a few places so problematic. 

backticks

```
const oldUIDQuery = {
   key:  `events#ba*`
};
```

missing quote

```
const wildcardQuery = {
    field: 'uuid,
    value: 'afe1*',
};
```

**Pull request merged**